README.md
# Server-to-server Library Inventory Management System
Computer Networks Socket Programming project

This repository contains the source code for a server system that manages book inventory using both UDP and TCP protocols written in C++.

The main server performs the following functions:

- Authenticates users against a member list.
- Listens for book availability requests over TCP.
- Queries backend servers for book statuses using UDP.
- Receives book status updates from backend servers.

## Environments

- Ubuntu 22.04 VM running on M1 Mac.
- Support for C++11 or higher.
- GNU Compiler Collection (GCC) or any compatible C++ compiler.
- Make (optional for compiling, can use direct compiler commands).


history.txt
H104, 9
H103, 5
H201, 7
H207, 2
H305, 1
H308, 4


client.cpp
//
// Created by wenxuan on 10/23/23.
//
#include <iostream>
#include <cstring>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

#define M_TCP_PORT 45986
#define M_IP "127.0.0.1"

std::string encrypt(const std::string &input) {
    std::string encrypted = "";
    for (char c : input) {
        if (isalpha(c)) {
            if (isupper(c))
                encrypted += (c - 'A' + 5) % 26 + 'A';
            else
                encrypted += (c - 'a' + 5) % 26 + 'a';
        } else if (isdigit(c)) {
            encrypted += (c - '0' + 5) % 10 + '0';
        } else {
            encrypted += c;
        }
    }
    return encrypted;
}

int main() {
    std::cout << "Client is up and running." << std::endl;
    
    int sockfd;
    struct sockaddr_in serverMAdd;
    char buffer[1024];
    
    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) {
        std::cerr << "Client socket creation error." << std::endl;
        exit(EXIT_FAILURE);
    }

    serverMAdd.sin_family = AF_INET;
    serverMAdd.sin_port = htons(M_TCP_PORT);
    inet_pton(AF_INET, M_IP, &serverMAdd.sin_addr);

    if (connect(sockfd, (struct sockaddr *)&serverMAdd, sizeof(serverMAdd)) < 0) {
        std::cerr << "TCP Connection to the server failed." << std::endl;
        exit(EXIT_FAILURE);
    }

    bool isAuthenticated = false;

    while (!isAuthenticated) {
        std::string username, password;
        std::cout << "Please enter the username: ";
        std::cin >> username;
        std::cout << "Please enter the password: ";
        std::cin >> password;

        std::string credential = encrypt(username) + "," + encrypt(password);
        send(sockfd, credential.c_str(), credential.length(), 0);
//        std::cout << credential << std::endl;
        std::cout << username << " sent an authentication request to the Main Server." << std::endl;

        int recvlen = recv(sockfd, buffer, sizeof(buffer), 0);
        buffer[recvlen] = '\0';

        std::string response(buffer);
        struct sockaddr_in localAddress;
        socklen_t addressLength = sizeof(localAddress);
        getsockname(sockfd, (struct sockaddr*)&localAddress, &addressLength);
        int clientPort = ntohs(localAddress.sin_port);

        if (response == "AUTH_SUCCESS") {
            std::cout << username <<  " received the result of authentication from Main Server using TCP over port "
                      << clientPort << ". Authentication is successful." << std::endl;
            isAuthenticated = true;

            //todo: phase III
            while (true) {
                std::cout << "Please enter book code to query: ";
                std::string bookCode;
                std::cin >> bookCode;
                send(sockfd, bookCode.c_str(), bookCode.length(), 0);
                std::cout << username << " sent the request to the Main Server." << std::endl;

                memset(buffer, 0, sizeof(buffer));
                recvlen = recv(sockfd, buffer, sizeof(buffer), 0);
                if (recvlen < 0) {
                    std::cerr << "Failed to receive data from Main server." << std::endl;
                    break;
                }
                buffer[recvlen] = '\0';
                std::string response(buffer);
                std::cout << "Response received from the Main Server on TCP port: " << clientPort << "." << std::endl;

                if (response == "BOOK_AVAILABLE") {
                    std::cout << "The requested book " << bookCode << " is available in the library." << std::endl;
                } else if (response == "BOOK_NOT_AVAILABLE") {
                    std::cout << "The requested book " << bookCode << " is NOT available in the library." << std::endl;
                } else {
                    std::cout << "Not able to find the book-code " << bookCode << " in the system." << std::endl;
                }
            }

        } else if (response == "USERNAME_NOT_FOUND") {
            std::cout << username << " received the result of authentication from Main Server using TCP over port "
                      << clientPort << ". Authentication failed: Username not found." << std::endl;
        } else if (response == "PASSWORD_NOT_MATCH") {
            std::cout << username << " received the result of authentication from Main Server using TCP over port "
                      << clientPort << ". Authentication failed: Password does not match." << std::endl;
        }
    }


    close(sockfd);
    return 0;
}


serverH.cpp
//
// Created by wenxuan on 10/23/23.
//

#include <iostream>
#include <fstream>
#include <sstream>
#include <map>
#include <cstring>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

#define H_PORT 43986
#define M_PORT 44986
#define M_IP "127.0.0.1"

std::map<std::string, int> historyStatus;

bool readStatus() {
    std::ifstream file("history.txt");
    if (!file.is_open()) {
        std::cerr << "No file history.txt found." << std::endl;
        return false;
    }

    std::string line;
    while (getline(file, line)) {
        std::stringstream ss(line);
        std::string bookCode;
        int count;
        getline(ss, bookCode, ',');
        ss >> count;
        historyStatus[bookCode] = count;
    }

//    for (const auto& [bookCode, count] : status) {
//        std::cout << "Book Code: " << bookCode << ", Count: " << count << std::endl;
//    }

    file.close();
    return true;
}

std::string processQuery(const std::string& bookCode) {
    if (historyStatus.find(bookCode) == historyStatus.end()) {
        return "BOOK_NOT_FOUND";
    } else if (historyStatus[bookCode] > 0) {
        historyStatus[bookCode]--;
        return std::to_string(historyStatus[bookCode] + 1);
    } else {
        return "0";
    }
}

bool sendToMainServer() {
    int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        std::cerr << "serverH UDP socket creation error." << std::endl;
        return false;
    }
    struct sockaddr_in serverMAdd;
    memset(&serverMAdd, 0, sizeof(serverMAdd));
    serverMAdd.sin_family = AF_INET;
    serverMAdd.sin_port = htons(M_PORT);
    inet_pton(AF_INET, M_IP, &serverMAdd.sin_addr);
    for (const auto& [bookCode, count] : historyStatus) {
        std::string message = bookCode + "," + std::to_string(count);
        sendto(sockfd, message.c_str(), message.size(), 0,
               (struct sockaddr *)&serverMAdd, sizeof(serverMAdd));
    }
    std::string terminator = "END_OF_LIST_H";
    sendto(sockfd, terminator.c_str(), terminator.size(), 0,
           (struct sockaddr *)&serverMAdd, sizeof(serverMAdd));
    close(sockfd);
    return true;
}

int main() {
    if (!readStatus()) {
        return 1;
    }

    int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        std::cerr << "serverH UDP socket creation error." << std::endl;
        return 1;
    }

    struct sockaddr_in serverHAddress;
    memset(&serverHAddress, 0, sizeof(serverHAddress));
    serverHAddress.sin_family = AF_INET;
    serverHAddress.sin_port = htons(H_PORT);
    serverHAddress.sin_addr.s_addr = INADDR_ANY;

    int bound = bind(sockfd, (struct sockaddr *)&serverHAddress, sizeof(serverHAddress));
    if (bound < 0) {
        std::cerr << "serverH UDP socket bind error." << std::endl;
        close(sockfd);
        return 1;
    }

    std::cout << "server H is up and running using UDP on port " << H_PORT << "." << std::endl;

    while (true) {
        char buffer[1024];
        struct sockaddr_in sourceAdd;
        socklen_t sourceAddLen = sizeof(sourceAdd);

        ssize_t recvBytes = recvfrom(sockfd, buffer, sizeof(buffer) - 1, 0,
                                     (struct sockaddr*)&sourceAdd, &sourceAddLen);
        if (recvBytes > 0) {
            buffer[recvBytes] = '\0';
            std::string bookCode(buffer);
            std::string response = processQuery(bookCode);

            std::cout << "Server H received " << bookCode << " code from the Main Server." << std::endl;

            sendto(sockfd, response.c_str(), response.length(), 0,
                   (struct sockaddr*)&sourceAdd, sizeof(sourceAdd));

            std::cout << "Server H finished sending the availability status of code " << bookCode <<
                      " to the Main Server using UDP on port " << H_PORT << "." << std::endl;
        }
    }
    
//    sendToMainServer();
    close(sockfd);
    return 0;
}

serverL.cpp
//
// Created by wenxuan on 10/23/23.
//

#include <iostream>
#include <fstream>
#include <sstream>
#include <map>
#include <cstring>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

#define L_PORT 42986
#define M_PORT 44986
#define M_IP "127.0.0.1"

std::map<std::string, int> literatureStatus;

bool readStatus() {
    std::ifstream file("literature.txt");
    if (!file.is_open()) {
        std::cerr << "No file literature.txt found." << std::endl;
        return false;
    }

    std::string line;
    while (getline(file, line)) {
        std::stringstream ss(line);
        std::string bookCode;
        int count;
        getline(ss, bookCode, ',');
        ss >> count;
        literatureStatus[bookCode] = count;
    }

//    for (const auto& [bookCode, count] : status) {
//        std::cout << "Book Code: " << bookCode << ", Count: " << count << std::endl;
//    }

    file.close();
    return true;
}

std::string processQuery(const std::string& bookCode) {
    if (literatureStatus.find(bookCode) == literatureStatus.end()) {
        return "BOOK_NOT_FOUND";
    } else if (literatureStatus[bookCode] > 0) {
        literatureStatus[bookCode]--;
        return std::to_string(literatureStatus[bookCode] + 1);
    } else {
        return "0";
    }
}

bool sendToMainServer() {
    int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        std::cerr << "serverL UDP socket creation error." << std::endl;
        return false;
    }
    struct sockaddr_in serverMAdd;
    memset(&serverMAdd, 0, sizeof(serverMAdd));
    serverMAdd.sin_family = AF_INET;
    serverMAdd.sin_port = htons(M_PORT);
    inet_pton(AF_INET, M_IP, &serverMAdd.sin_addr);
    for (const auto& [bookCode, count] : literatureStatus) {
        std::string message = bookCode + "," + std::to_string(count);
        sendto(sockfd, message.c_str(), message.size(), 0,
               (struct sockaddr *)&serverMAdd, sizeof(serverMAdd));
    }
    std::string terminator = "END_OF_LIST_L";
    sendto(sockfd, terminator.c_str(), terminator.size(), 0,
           (struct sockaddr *)&serverMAdd, sizeof(serverMAdd));
    close(sockfd);
    return true;
}

int main() {
    if (!readStatus()) {
        return 1;
    }

    int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        std::cerr << "serverL UDP socket creation error." << std::endl;
        return 1;
    }

    struct sockaddr_in serverLAddress;
    memset(&serverLAddress, 0, sizeof(serverLAddress));
    serverLAddress.sin_family = AF_INET;
    serverLAddress.sin_port = htons(L_PORT);
    serverLAddress.sin_addr.s_addr = INADDR_ANY;

    int bound = bind(sockfd, (struct sockaddr *)&serverLAddress, sizeof(serverLAddress));
    if (bound < 0) {
        std::cerr << "serverL UDP socket bind error." << std::endl;
        close(sockfd);
        return 1;
    }

    std::cout << "Server L is up and running using UDP on port " << L_PORT << "." << std::endl;

    while (true) {
        char buffer[1024];
        struct sockaddr_in sourceAdd;
        socklen_t sourceAddLen = sizeof(sourceAdd);

        ssize_t recvBytes = recvfrom(sockfd, buffer, sizeof(buffer) - 1, 0,
                                     (struct sockaddr*)&sourceAdd, &sourceAddLen);
        if (recvBytes > 0) {
            buffer[recvBytes] = '\0';
            std::string bookCode(buffer);
            std::string response = processQuery(bookCode);

            std::cout << "Server L received " << bookCode << " code from the Main Server." << std::endl;

            sendto(sockfd, response.c_str(), response.length(), 0,
                   (struct sockaddr*)&sourceAdd, sizeof(sourceAdd));

            std::cout << "Server L finished sending the availability status of code " << bookCode <<
                      " to the Main Server using UDP on port " << L_PORT << "." << std::endl;
        }
    }

//    sendToMainServer();
    close(sockfd);
    return 0;
}

serverS.cpp
//
// Created by wenxuan on 10/18/23.
//

#include <iostream>
#include <fstream>
#include <sstream>
#include <map>
#include <cstring>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

#define S_PORT 41986
#define M_PORT 44986
#define M_IP "127.0.0.1"

std::map<std::string, int> scienceStatus;

bool readStatus() {
    std::ifstream file("science.txt");
    if (!file.is_open()) {
        std::cerr << "No file science.txt found." << std::endl;
        return false;
    }

    std::string line;
    while (getline(file, line)) {
        std::stringstream ss(line);
        std::string bookCode;
        int count;
        getline(ss, bookCode, ',');
        ss >> count;
        scienceStatus[bookCode] = count;
    }

//    for (const auto& [bookCode, count] : status) {
//        std::cout << "Book Code: " << bookCode << ", Count: " << count << std::endl;
//    }

    file.close();
    return true;
}

std::string processQuery(const std::string& bookCode) {
    if (scienceStatus.find(bookCode) == scienceStatus.end()) {
        return "BOOK_NOT_FOUND";
    } else if (scienceStatus[bookCode] > 0) {
        scienceStatus[bookCode]--;
        return std::to_string(scienceStatus[bookCode] + 1);
    } else {
        return "0";
    }
}

bool sendToMainServer() {
    int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        std::cerr << "ServerS UDP socket creation error." << std::endl;
        return false;
    }
    struct sockaddr_in serverMAdd;
    memset(&serverMAdd, 0, sizeof(serverMAdd));
    serverMAdd.sin_family = AF_INET;
    serverMAdd.sin_port = htons(M_PORT);
    inet_pton(AF_INET, M_IP, &serverMAdd.sin_addr);
    for (const auto& [bookCode, count] : scienceStatus) {
        std::string message = bookCode + "," + std::to_string(count);
        sendto(sockfd, message.c_str(), message.size(), 0,
               (struct sockaddr *)&serverMAdd, sizeof(serverMAdd));
    }
    std::string terminator = "END_OF_LIST_S";
    sendto(sockfd, terminator.c_str(), terminator.size(), 0,
           (struct sockaddr *)&serverMAdd, sizeof(serverMAdd));
    close(sockfd);
    return true;
}

int main() {
    if (!readStatus()) {
        return 1;
    }

    int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        std::cerr << "ServerS UDP socket creation error." << std::endl;
        return 1;
    }

    struct sockaddr_in serverSAddress;
    memset(&serverSAddress, 0, sizeof(serverSAddress));
    serverSAddress.sin_family = AF_INET;
    serverSAddress.sin_port = htons(S_PORT);
    serverSAddress.sin_addr.s_addr = INADDR_ANY;

    int bound = bind(sockfd, (struct sockaddr *)&serverSAddress, sizeof(serverSAddress));
    if (bound < 0) {
        std::cerr << "ServerS UDP socket bind error." << std::endl;
        close(sockfd);
        return 1;
    }

    std::cout << "Server S is up and running using UDP on port " << S_PORT << "." << std::endl;

//    sendToMainServer();

    while (true) {
        char buffer[1024];
        struct sockaddr_in sourceAdd;
        socklen_t sourceAddLen = sizeof(sourceAdd);

        ssize_t recvBytes = recvfrom(sockfd, buffer, sizeof(buffer) - 1, 0,
                                     (struct sockaddr*)&sourceAdd, &sourceAddLen);
        if (recvBytes > 0) {
            buffer[recvBytes] = '\0';
            std::string bookCode(buffer);
            std::string response = processQuery(bookCode);

            std::cout << "Server S received " << bookCode << " code from the Main Server." << std::endl;

            sendto(sockfd, response.c_str(), response.length(), 0,
                   (struct sockaddr*)&sourceAdd, sizeof(sourceAdd));

            std::cout << "Server S finished sending the availability status of code " << bookCode <<
            " to the Main Server using UDP on port " << S_PORT << "." << std::endl;
        }
    }

    close(sockfd);
    return 0;
}

serverM.cpp
//
// Created by wenxuan on 10/22/23.
//
#include <iostream>
#include <cstring>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <map>
#include <unistd.h>
#include <fstream>
#include <thread>

#define M_UDP_PORT 44986
#define M_TCP_PORT 45986
#define S_UDP_PORT 41986
#define L_UDP_PORT 42986
#define H_UDP_PORT 43986

std::map<std::string, int> booksStatus;
std::map<std::string, std::string> members;
int udpSocket;

struct sockaddr_in serverSAdd, serverLAdd, serverHAdd;

void setupServerAddresses() {
    memset(&serverSAdd, 0, sizeof(serverSAdd));
    memset(&serverLAdd, 0, sizeof(serverLAdd));
    memset(&serverHAdd, 0, sizeof(serverHAdd));

    serverSAdd.sin_family = AF_INET;
    serverSAdd.sin_port = htons(S_UDP_PORT);

    serverLAdd.sin_family = AF_INET;
    serverLAdd.sin_port = htons(L_UDP_PORT);

    serverHAdd.sin_family = AF_INET;
    serverHAdd.sin_port = htons(H_UDP_PORT);

    inet_pton(AF_INET, "127.0.0.1", &serverSAdd.sin_addr);
    inet_pton(AF_INET, "127.0.0.1", &serverLAdd.sin_addr);
    inet_pton(AF_INET, "127.0.0.1", &serverHAdd.sin_addr);
}

void udpSendToBackend(int serverPort, const std::string& bookCode) {
    struct sockaddr_in serverAdd;
    memset(&serverAdd, 0, sizeof(serverAdd));
    serverAdd.sin_family = AF_INET;
    serverAdd.sin_port = htons(serverPort);
    serverAdd.sin_addr.s_addr = inet_addr("127.0.0.1");

    if (sendto(udpSocket, bookCode.c_str(), bookCode.size(), 0,
               (struct  sockaddr*)&serverAdd, sizeof(serverAdd)) < 0) {
        std::cerr << "Error sending bookCode to backend server." << std::endl;
    }
}

std::string udpReceiveFromBackend(int serverPort) {
    struct sockaddr_in fromAdd;
    socklen_t fromAddLen = sizeof(fromAdd);
    char buffer[1024];

    ssize_t recvBytes = recvfrom(udpSocket, buffer, sizeof(buffer) - 1, 0,
                                 (struct sockaddr*)&fromAdd, &fromAddLen);
    if (recvBytes < 0) {
        std::cerr << "Error receiving from server." << std::endl;
        return "";
    }

    buffer[recvBytes] = '\0';

    if (ntohs(fromAdd.sin_port) != serverPort) {
        std::cerr << "Received Book Status from an unexpected port." << std::endl;
        return "";
    }

    return std::string(buffer);
}

void checkBookAvailability(int childsockfd, const std::string& bookCode) {
    std::cout << "Main Server received the book request from client using TCP over port " << M_TCP_PORT << "." << std::endl;
    char serverHead = toupper(bookCode[0]);
    int serverPort;
    std::string serverString;
    switch (serverHead) {
        case 'S':
            serverPort = S_UDP_PORT;
            serverString = "S";
            break;
        case 'L':
            serverPort = L_UDP_PORT;
            serverString = "L";
            break;
        case 'H':
            serverPort = H_UDP_PORT;
            serverString = "H";
            break;
        default:
            std::cout << "Did not find " << bookCode << " in the book code list." << std::endl;
            send(childsockfd, "Book code not found.", 19, 0);
            return;
    }

    std::cout << "Found " << bookCode << " located at Server " << serverString << ". Send to Server " << serverString << "." << std::endl;
    udpSendToBackend(serverPort, bookCode); // Todo

    std::string bookStatus = udpReceiveFromBackend(serverPort); // Todo
    std::cout << "Main Server received from server " << serverString << " the book status result using UDP over port "
              << M_UDP_PORT << ":" << std::endl
              << "Number of books " << bookCode << " available is: " << bookStatus << "." << std::endl;

    if (bookStatus != "BOOK_NOT_FOUND") {
        if (bookStatus == "0") {
            bookStatus = "BOOK_NOT_AVAILABLE";
        } else {
            bookStatus = "BOOK_AVAILABLE";
        }
    }
    send(childsockfd, bookStatus.c_str(), bookStatus.length(), 0);
    std::cout << "Main Server sent the book status to the client." << std::endl;
}

void udpListen() {
    int sockfd;
    struct sockaddr_in serverAdd, clientAdd;
    char buffer[1024];

    sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        std::cerr << "ServerM UDP socket creation error." << std::endl;
        exit(EXIT_FAILURE);
    }

    memset(&serverAdd, 0, sizeof(serverAdd));
    serverAdd.sin_family = AF_INET;
    serverAdd.sin_addr.s_addr = INADDR_ANY;
    serverAdd.sin_port = htons(M_UDP_PORT);

    if (bind(sockfd, (struct sockaddr *) &serverAdd, sizeof(serverAdd)) < 0) {
        std::cerr << "ServerM UDP socket binding error." << std::endl;
        exit(EXIT_FAILURE);
    }

    while (true) {
        socklen_t clientLen = sizeof(clientAdd);
        int recvlen = recvfrom(sockfd, buffer, sizeof(buffer), 0, (struct sockaddr *) &clientAdd, &clientLen);
        if (recvlen > 0) {
            buffer[recvlen] = 0;
            std::string data(buffer);
            if (data.find("END_OF_LIST") == 0) {
                char serverNum = data[data.length() - 1];
                std::cout << "Main Server received the book code list from server" << serverNum
                          << " using UDP over port " << M_UDP_PORT << "." << std::endl;
//                for (const auto& [bookCode, count] : booksStatus) {
//                    std::cout << "Book Code: " << bookCode << ", Count: " << count << std::endl;
//                }
                continue;
            }
            size_t splitter = data.find(",");
            std::string bookCode = data.substr(0, splitter);
            int status = std::stoi(data.substr(splitter + 1));
            booksStatus[bookCode] = status;
        } else if (recvlen < 0) {
            std::cerr << "ServerM recvfrom error." << std::endl;
        }
    }

    close(sockfd);
}

void readMembers() {
    std::ifstream inFile("member.txt");
    std::string username, password;
    while (inFile >> username >> password) {
        username.pop_back();
        members[username] = password;
    }
    inFile.close();
    std::cout << "Main Server loaded the member list." << std::endl;
}

bool authenticateUser(int childsockfd, const std::string &input) {
    size_t splitter = input.find(",");
    std::string username = input.substr(0, splitter);
    std::string password = input.substr(splitter + 1);
    std::cout << "Main Server received the username " << username
            << " and password " << password << " from the client using TCP over port "
            << M_TCP_PORT << "." << std::endl;

    std::string response;
    if (members.find(username) == members.end()) {
        response = "USERNAME_NOT_FOUND";
        std::cout << username << " is not registered. Send a reply to the client." << std::endl;
    } else if (members[username] == password) {
        response = "AUTH_SUCCESS";
        std::cout << "Password " << password << " matches the username. Send a reply to the client." << std::endl;
    } else {
        response = "PASSWORD_NOT_MATCH";
        std::cout << "Password " << password << " does not match the username. Send a reply to the client." << std::endl;
    }
    send(childsockfd, response.c_str(), response.length(), 0);

    return response == "AUTH_SUCCESS";
}

void tcpListen() {
    int sockfd, childsockfd;
    struct sockaddr_in serverMAdd, clientAdd;
    char buffer[1024];
    
    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) {
        std::cerr << "ServerM TCP socket creation error." << std::endl;
        exit(EXIT_FAILURE);
    }

    memset(&serverMAdd, 0, sizeof(serverMAdd));
    serverMAdd.sin_family = AF_INET;
    serverMAdd.sin_addr.s_addr = INADDR_ANY;
    serverMAdd.sin_port = htons(M_TCP_PORT);

    if (bind(sockfd, (struct sockaddr*)&serverMAdd, sizeof(serverMAdd)) < 0) {
        std::cerr << "ServerM TCP socket binding error." << std::endl;
        close(sockfd);
        exit(EXIT_FAILURE);
    }

    if (listen(sockfd, 3) < 0) {
        std::cerr << "ServerM TCP socket listening error." << std::endl;
        close(sockfd);
        exit(EXIT_FAILURE);
    }

//    std::cout << "TCP Server listening on port " << M_TCP_PORT << std::endl;

    socklen_t clientLen = sizeof(clientAdd);

    while (true) {
        childsockfd = accept(sockfd, (struct sockaddr*)&clientAdd, &clientLen);
        if (childsockfd < 0) {
            std::cerr << "Error accepting TCP connection." << std::endl;
            continue;
        }

        bool isAuthenticated = false;

        while (!isAuthenticated) {
            int recvlen = read(childsockfd, buffer, 1024);
            if (recvlen < 0) {
                std::cerr << "Error reading from TCP." << std::endl;
                break;
            } else if (recvlen == 0) {
                close(childsockfd);
                break;
            } else {
                buffer[recvlen] = '\0';
                isAuthenticated = authenticateUser(childsockfd, buffer);
            }
        }

        while (isAuthenticated) {
            int recvlen = read(childsockfd, buffer, 1024);
            if (recvlen < 0) {
                std::cerr << "Error reading from TCP." << std::endl;
                close(childsockfd);
                isAuthenticated = false;
            } else if (recvlen == 0) {
                close(childsockfd);
            } else {
                buffer[recvlen] = '\0';
                std::string bookCode(buffer);

                checkBookAvailability(childsockfd, bookCode); // todo
            }
        }

    }

    close(childsockfd);
}

int main() {
    std::cout << "Main Server is up and running." << std::endl;
    readMembers();
//    for (const auto& [username, passward] : members) {
//        std::cout << "username: " << username << ", password: " << passward << std::endl;
//    }
    setupServerAddresses();
    udpSocket = socket(AF_INET, SOCK_DGRAM, 0);
    if (udpSocket < 0) {
        std::cerr << "Error creating UDP socket." << std::endl;
        return 1;
    }
    struct sockaddr_in mainServerAddr;
    memset(&mainServerAddr, 0, sizeof(mainServerAddr));
    mainServerAddr.sin_family = AF_INET;
    mainServerAddr.sin_port = htons(M_UDP_PORT);
    mainServerAddr.sin_addr.s_addr = htonl(INADDR_ANY);

    if (bind(udpSocket, (struct sockaddr *)&mainServerAddr, sizeof(mainServerAddr)) < 0) {
        std::cerr << "Error binding UDP socket to port." << std::endl;
        close(udpSocket);
        return 1;
    }
//    std::thread udpThread(udpListen);
    std::thread tcpThread(tcpListen);

//    udpThread.join();
    tcpThread.join();

    close(udpSocket);

    return 0;
}
